jobs/integration/validation.py

```diff
@@ -1372,19 +1373,56 @@ def _find_relation(*specs):
     elif len(keystone_apps) == 1:
         # One keystone found, ensure related to kubernetes-master
         keystone, *_ = keystone_apps.values()
-        credentials_rel = _find_relation(keystone_creds)
-        if credentials_rel is None:
+        credentials_rel = list(_find_relation(keystone_creds))
+        if not credentials_rel:
             await keystone.add_relation("identity-credentials", keystone_creds)
             await tools.juju_wait()
 
         keystone_master = random.choice(keystone.units)
         action = await keystone_master.run("leader-get admin_passwd")
         admin_password = action.results.get("Stdout", "").strip()
+
+        # is the same vault supplying tls-certificates for keystone, kubernetes-master, and kubernetes-worker
+        keystone_ssl_ca = (await masters.get_config())["keystone-ssl-ca"]["value"]
+        if not keystone_ssl_ca:
+            vault_root_ca = None
+            vault_apps = apps_by_charm("vault")
+            for name, vault_app in vault_apps.items():
+                vault_tls = f"{name}:certificates"
+                rels = set(
+                    app.name
+                    for rel in _find_relation(vault_tls)
+                    for app in rel.applications
+                )
+                if all(
+                    name in rels
+                    for name in [
+                        "kubernetes-master",
+                        "kubernetes-worker",
+                        keystone.name,
+                    ]
+                ):
+                    vault_unit = random.choice(vault_app.units)
+                    action = await vault_unit.run_action("get-root-ca")
+                    await action.wait()
+                    assert action.status not in ("pending", "running", "failed")
+                    vault_root_ca = action.results.get("output")
+                    if vault_root_ca:
+                        vault_root_ca = base64.b64encode(
+                            vault_root_ca.encode("ascii")
+                        ).decode("ascii")
+                        break
+
+            if vault_root_ca:
```

Comment:
I think this is worth noting that it is a workaround for a bug 
It should make finding the code to remove when the bug is fixed easier as well